eth_defi/aave_v3/rates.py

```diff
@@ -0,0 +1,174 @@
+    if isinstance(attribute, str):
+        # Single attribute
+        return df[attribute].resample(time_bucket).ohlc(_method='ohlc')
+    else:
+        # Multiple attributes
+        return (df[attr].resample(time_bucket).ohlc(_method='ohlc') for attr in attribute)
+
+
+def aave_v3_calculate_mean(df: DataFrame, time_bucket: Timedelta, attribute: str | Tuple, token: str = '') -> DataFrame | Tuple:
+    """
+    Calculate mean values for a given time bucket (e.g. 1 day) and given attribute.
+    Attribute can be e.g. deposit_apr, variable_borrow_apr, stable_borrow_apr, deposit_apy, variable_borrow_apy, stable_borrow_apy.
+    The dataframe must be indexed by timestamp.
+    Returns a new DataFrame, or a tuple of DataFrames if a tuple of attributes was specified.
+    """
+    df = aave_v3_filter_by_token(df, token)
+    if isinstance(attribute, str):
+        # Single attribute
+        return df[attribute].resample(time_bucket).mean()
+    else:
+        # Multiple attributes
+        return (df[attr].resample(time_bucket).mean() for attr in attribute)
+
+
+def aave_v3_filter_by_date_range(df: DataFrame, start_time: datetime, end_time: datetime = None, token: str = '') -> DataFrame:
+    """
+    Filter the DataFrame by date range suitable for interest calculation (loan start to loan end time)
+    The DataFrame must be indexed by timestamp.
+    If token is specified, also filters by token.
+    """
+    if end_time:
+        return aave_v3_filter_by_token(df, token).query('timestamp >= @start_time and timestamp <= @end_time')
+    else:
+        return aave_v3_filter_by_token(df, token).query('timestamp >= @start_time')
+
+
+def _calculate_compound_interest_multiplier(rate: Decimal, seconds: Decimal | float) -> Decimal:
+    """
+    Calculate compound interest for a given rate and seconds between borrow and payback.
+    Based on https://github.com/aave/aave-v3-core/blob/master/contracts/protocol/libraries/math/MathUtils.sol#L51
+    """
+    exp = Decimal(seconds)  # ensure we use decimal
+    if exp == 0:
+        # No time elapsed, no interest accrued.
+        return 1
+    exp_minus_one = exp - Decimal(1)
+    exp_minus_two = (exp - Decimal(2)) if exp > 2 else Decimal(0)
+    base_power_two = rate * rate / (SECONDS_PER_YEAR * SECONDS_PER_YEAR)
+    base_power_three = base_power_two * rate / SECONDS_PER_YEAR
+    second_term = exp * exp_minus_one * base_power_two / Decimal(2)
+    third_term = exp * exp_minus_one * exp_minus_two * base_power_three / Decimal(6)
+    multiplier = rate * exp / SECONDS_PER_YEAR + second_term + third_term + 1
+    # Alternative simplified Python implementation
+    # multiplier = ((rate / SECONDS_PER_YEAR) + 1) ** exp
+    # logger.warn(f'Multiplier: {multiplier}')
+    return multiplier
+
+
+def aave_v3_calculate_accrued_interests(df: DataFrame, start_time: datetime, end_time: datetime, amount: Decimal, token: str = '') -> Tuple[datetime, datetime, Decimal, Decimal, Decimal]:
```

Comment:
Since the resulting tuple contains quite a few fields, I think it would be a good idea to define a dedicated namedtuple for it. It will make the code that uses this function easier to read.

E.g. `return result[0], result[1], result[4]` is considerably less  clear than if named fields were used.

Reply:
Definitely makes sense to me, I'll add it.